File: mnlc/mnlc_navigation/mnlc_global_navigation/mnlc_a_star.py
#!/usr/bin/env python

from nav_msgs.msg import OccupancyGrid, Path, GridCells
from geometry_msgs.msg import PoseStamped, Point
from nav_msgs.srv import GetPlan, GetMap
from queues import PriorityQueueDep
from scipy.spatial import distance
import move_base_msgs.msg as mb
import std_srvs.srv
import numpy as np
import actionlib
import roslib
import rospy
import math

roslib.load_manifest('rbe3002')


class mnlc_a_star():

    def __init__(self):
        self.error = False
        rospy.loginfo("Initializing mnlc_a_star.")
        rospy.init_node("mnlc_a_star")
        self.initialize_params()
        rospy.sleep(self.start_time)
        self.safe_start()
        # give gazebo a chance to warm up so rtabmap doesnt raise an error about not having a map
        rospy.loginfo("mnlc_a_star node ready.")

    def initialize_params(self):
        self.start_time = rospy.get_param('/a_star/start_time')
        # grid cost to be considered an obstacle
        self.obstacle_cost = rospy.get_param('/a_star/obstacle_cost')
        # [s] standard service timeout limit
        self.timeout = rospy.get_param('/controller/timeout')
        self.local_costmap = OccupancyGrid()
        self.global_costmap = OccupancyGrid()
        self.update_global_costmap_sub = rospy.Subscriber(
            '/mnlc_global_costmap_opencv/cspace', OccupancyGrid, self.update_global_costmap, queue_size=1)
        self.frontier_pub = rospy.Publisher(
            '/a_star/frontier', GridCells, queue_size=1)

    def safe_start(self):
        rospy.wait_for_service(
            '/rtabmap/get_map', timeout=rospy.Duration(self.timeout))
        try:
            m = rospy.ServiceProxy('/rtabmap/get_map', GetMap)
        except rospy.ServiceException as e:
            rospy.logerr("RTabMap Mappping service call failed: %s" % e)
            self.error_handler()
            return
        rospy.loginfo("RTabMap Mappping service call successful.")
        map = m()
        if map.map.info.resolution == 0:
            rospy.logerr(
                "RTabMap has zero resolution. Exiting...")
            self.error_handler()
            return
        rospy.wait_for_service(
            '/begin_phase1', timeout=rospy.Duration(self.timeout))
        try:
            tmp = rospy.ServiceProxy('/begin_phase1', std_srvs.srv.Empty)
        except rospy.ServiceException as e:
            rospy.logerr("Begin phase1 service call failed: %s" % e)
            self.error_handler()
            return
        rospy.loginfo("Begin phase1 service call successful.")
        temp = tmp()
        rospy.Service('/plan_path', GetPlan, self.plan_path)

    def plan_path(self, msg):
        rospy.loginfo("A* is planning the path.")
        res = self.global_costmap.info.resolution
        ox = self.global_costmap.info.origin.position.x
        oy = self.global_costmap.info.origin.position.y
        x1 = int(math.floor((msg.start.pose.position.x - ox) / res))
        y1 = int(math.floor((msg.start.pose.position.y - oy) / res))
        start = (x1, y1)
        x2 = int(math.floor((msg.goal.pose.position.x - ox) / res))
        y2 = int(math.floor((msg.goal.pose.position.y - oy) / res))
        goal = (x2, y2)
        path = self.a_star(start, goal)
        path_msg = Path()
        path_msg.header.frame_id = '/map'
        if path == -1:
            empty_pose = PoseStamped()
            empty_pose.header.frame_id = '/map'
            empty_pose.header.stamp = rospy.Time.now()
            path_msg.poses = [empty_pose]
            path_msg.header.stamp = rospy.Time.now()
            rospy.loginfo(
                "A* has not found a path within the alotted time. Send a new goal point.")
            return path_msg
        for point in path:
            pose = PoseStamped()
            point_ret = Point()
            point_ret.x = (point[0] * res) + ox + (res/2)
            point_ret.y = (point[1] * res) + oy + (res/2)
            point_ret.z = 0
            pose.pose.position.x = point_ret.x
            pose.pose.position.y = point_ret.y
            pose.header.stamp = rospy.Time.now()
            pose.header.frame_id = '/map'
            path_msg.poses.append(pose)
        path_msg.header.stamp = rospy.Time.now()
        rospy.loginfo("A* has computed the path.")
        return path_msg

    def a_star(self, start, goal):
        rospy.loginfo("Beginning A* algorithm.")
        time_init = rospy.get_time()
        rospy.loginfo("Executing A* from (%d,%d) to (%d,%d)." %
                      (start[0], start[1], goal[0], goal[1]))
        frontier_nodes = GridCells()
        obstacle_cost = self.obstacle_cost
        frontier = PriorityQueueDep()
        frontier.put(start, 0)
        goal_np = np.array(goal)
        came_from = dict()
        cost_so_far = dict()
        came_from[start] = None
        cost_so_far[start] = 0
        data = self.global_costmap.data
        width = self.global_costmap.info.width
        height = self.global_costmap.info.height
        res = self.global_costmap.info.resolution
        ox = self.global_costmap.info.origin.position.x
        oy = self.global_costmap.info.origin.position.y
        frontier_nodes.cell_height = self.global_costmap.info.resolution
        frontier_nodes.cell_width = self.global_costmap.info.resolution
        start_time = rospy.get_time()
        while not frontier.empty():
            current = frontier.get()
            if current == goal:
                rospy.loginfo("Path has been found.")
                break
            time = rospy.get_time()
            if time >= start_time + 10.0:
                rospy.logwarn("A* has not found a path in less than 10.0 seconds. Requesting new path...")
                return -1
            x = current[0]
            y = current[1]
            cell_neighbors = [(x + 1, y + 1), (x + 1, y), (x + 1, y - 1), (x, y + 1),
                              (x, y - 1), (x - 1, y + 1), (x - 1, y), (x - 1, y - 1)]
            valid_neighbors = [cell for cell in cell_neighbors if ((cell[0] > 0) and (cell[0] < width) and (
                cell[1] > 0) and (cell[1] < height) and (data[cell[0] + (cell[1] * width)] < obstacle_cost))]
            for next in valid_neighbors:
                c = data[next[0] + (next[1] * width)]
                if c == -1:
                    c = 25
                new_cost = cost_so_far[current] + c
                if next not in cost_so_far or new_cost < cost_so_far[next]:
                    cost_so_far[next] = new_cost
                    next_np = np.array(next)
                    priority = new_cost + 4.0 * distance.euclidean(goal_np, next_np)
                    frontier.put(next, priority)
                    came_from[next] = current
                    point = Point()
                    point.x = (x * res) + ox + res/2
                    point.y = (y * res) + oy + res/2
                    frontier_nodes.cells.append(point)
        frontier_nodes.cells.reverse()
        frontier_nodes.header.frame_id = "map"
        frontier_nodes.header.stamp = rospy.Time.now()
        self.frontier_pub.publish(frontier_nodes)
        path = self.reconstruct_path(came_from, start, goal)
        rospy.loginfo("A* has successfully found the optimal path.")
        print("Calculating A* took: ", rospy.get_time() - time_init, ".")
        return path

    def reconstruct_path(self, came_from, start, goal):
        rospy.loginfo("Reconstructing path.")
        current = goal
        path = []
        while current != start:
            path.append(current)
            current = came_from[current]
        path.append(start)
        path.reverse()
        rospy.loginfo("Path reconstructed.")
        return path

    def safe_start_phase_2(self):
        rospy.wait_for_service(
            '/round_up_nodes', timeout=rospy.Duration(self.timeout))
        try:
            tmp = rospy.ServiceProxy('/round_up_nodes', std_srvs.srv.Empty)
        except rospy.ServiceException as e:
            rospy.logerr("Round up nodes service call failed: %s" % e)
            self.error_handler()
            return
        rospy.loginfo("Round up nodes service call successful.")
        temp = tmp()
        rospy.wait_for_service(
            '/begin_phase2', timeout=rospy.Duration(self.timeout))
        try:
            tmp2 = rospy.ServiceProxy('/begin_phase2', std_srvs.srv.Empty)
        except rospy.ServiceException as e:
            rospy.logerr("Begin phase2 service call failed: %s" % e)
            self.error_handler()
            return
        rospy.loginfo("Begin phase2 service call successful.")
        temp2 = tmp2()
        self.phase2_client = actionlib.SimpleActionClient(
            '/mnlc/navigate_to_origin', mb.MoveBaseAction)

    def final(self):
        self.phase3_client = actionlib.SimpleActionClient(
            '/mnlc/navigation', mb.MoveBaseAction)
        self.phase3_client.wait_for_server(
            timeout=rospy.Duration(self.timeout))

    def update_global_costmap(self, map):
        if map.info.resolution != 0.0:
            self.global_costmap = map
        else:
            rospy.logwarn("Receiving a global costmap with 0 resolution!")

    def update_local_costmap(self, map):
        if map.info.resolution != 0.0:
            self.local_costmap = map
        else:
            rospy.logwarn("Receiving a local costmap with 0 resolution!")

    def error_handler(self):
        self.error = True

    def run(self):
        while not rospy.is_shutdown() and self.error == False:
            rospy.spin()


if __name__ == '__main__':
    try:
        mnlc_a_star().run()
    except rospy.ROSInterruptException:
        pass

File: src/mnlc_a_star.cpp
#include "mnlc_a_star.h"

void a_star(Node start, Node goal, std::vector<geometry_msgs::Point> &path)
{
    double time_init = ros::Time(0).toSec();
    ROS_INFO("Beginning A* algorithm");
    printf("Executing A* from (%d, %d) to (%d, %d).", start.x, start.y, goal.x, goal.y);
    int width = mapdata.info.width;
    int height = mapdata.info.height;
    float res = mapdata.info.resolution;
    float gox = mapdata.info.origin.position.x;
    float goy = mapdata.info.origin.position.y;
    std::vector<int8_t, std::allocator<int8_t>> data = mapdata.data;
    frontier_vis.cells.clear();
    frontier_vis.cell_height = frontier_vis.cell_width = res;
    frontier_vis.header.frame_id = "/map";
    std::priority_queue<Node, std::vector<Node>, std::function<bool(Node, Node)>> frontiers(Compare);
    frontiers.push(start);
    std::unordered_map<int, int> came_from;
    std::unordered_map<int, int> cost_so_far;
    int start_index = start.x + (start.y * width);
    int goal_index = goal.x + (goal.y * width);
    came_from[start_index] = NULL;
    cost_so_far[start_index] = 0;
    //   for (auto itr = came_from.begin(); itr != came_from.end(); itr++)
    //     std::cout << itr->first << "\t" << itr->second << std::endl;
    // goes ccw starting from middle right node
    Node nbr1(goal.x + 1, goal.y + 0);
    Node nbr2(goal.x + 1, goal.y + 1);
    Node nbr3(goal.x + 0, goal.y + 1);
    Node nbr4(goal.x - 1, goal.y + 1);
    Node nbr5(goal.x - 1, goal.y + 0);
    Node nbr6(goal.x - 1, goal.y - 1);
    Node nbr7(goal.x + 0, goal.y - 1);
    Node nbr8(goal.x + 1, goal.y - 1);
    Node gnbrs[9] = {goal, nbr1, nbr2, nbr3, nbr4, nbr5, nbr6, nbr7, nbr8};
    for (int i = 0; i < 9; i++)
    {
        if (data[gnbrs[i].x + (gnbrs[i].y * width)] >= obstacle_cost)
        {
            ROS_WARN("A* has determined that the goal is invalid. Requesting new path...");
            return;
        }
    }
    Node connbrs[8];
    double start_time = ros::Time(0).toSec();
    while (frontiers.empty() == false)
    {
        Node current = frontiers.top();
        frontiers.pop();
        int cindex = current.x + (current.y * width);
        if ((current.x == goal.x) && (current.y == goal.y))
        {
            ROS_INFO("Path has been found.");
            break;
        }
        if (ros::Time(0).toSec() >= start_time + 1.0)
        {
            ROS_WARN("A* has not found a path in less than 1.0 seconds. Requesting new path...");
            return;
        }
        Node nbr1(current.x + 1, current.y + 0);
        connbrs[0] = nbr1;
        Node nbr2(current.x + 1, current.y + 1);
        connbrs[1] = nbr2;
        Node nbr3(current.x + 0, current.y + 1);
        connbrs[2] = nbr3;
        Node nbr4(current.x - 1, current.y + 1);
        connbrs[3] = nbr4;
        Node nbr5(current.x - 1, current.y + 0);
        connbrs[4] = nbr5;
        Node nbr6(current.x - 1, current.y - 1);
        connbrs[5] = nbr6;
        Node nbr7(current.x + 0, current.y - 1);
        connbrs[6] = nbr7;
        Node nbr8(current.x + 1, current.y - 1);
        connbrs[7] = nbr8;
        for (int i = 0; i < 8; i++)
        {
            int cost = data[connbrs[i].x + (connbrs[i].y * width)];
            int mcost;
            if (cost < 0)
            {
                mcost = 20;
            }
            else
            {
                mcost = cost;
            }
            if ((connbrs[i].x > 0) && (connbrs[i].x < width) &&
                (connbrs[i].y > 0) && (connbrs[i].y < height) &&
                (mcost < obstacle_cost))
            {
                int index = connbrs[i].x + (connbrs[i].y * width);
                int new_cost = cost_so_far[cindex] + mcost;
                if ((cost_so_far.find(index) == cost_so_far.end()) || (new_cost < cost_so_far[index]))
                {
                    cost_so_far[index] = new_cost;
                    connbrs[i].priority = new_cost + (0.5 * std::sqrt(std::pow(goal.x - connbrs[i].x, 2) + std::pow(goal.y - connbrs[i].y, 2)));
                    frontiers.push(connbrs[i]);
                    came_from[index] = cindex;
                    geometry_msgs::Point point;
                    point.x = (current.x * res) + gox + (res / 2);
                    point.y = (current.y * res) + goy + (res / 2);
                    frontier_vis.cells.push_back(point);
                }
            }
            // if (frontiers.empty()) {
            //     printf("Empty.\n");
            // } else {
            //     std::size_t size = frontiers.size();
            //     printf("Size: %zu\n", size);
            // }
        }
    }
    frontier_vis.header.stamp = ros::Time(0);
    ROS_INFO("Reconstructing path.");
    int c = goal_index;
    geometry_msgs::Point point;
    while (c != start_index)
    {
        point.x = ((c % width) * res) + gox + (res / 2);
        point.y = ((c / width) * res) + goy + (res / 2);
        path.push_back(point);
        // printf("c: %d\t", c);
        c = came_from[c];
        // printf("came from: %d\n", c);
    }
    point.x = ((start_index % width) * res) + gox + (res / 2);
    point.y = ((start_index / width) * res) + goy + (res / 2);
    path.push_back(point);
    ROS_INFO("Path reconstructed.");
    std::reverse(path.begin(), path.end());
    ROS_INFO("A* has successfully found the optimal path.");
    printf("Calculating A* took: %f\n", ros::Time(0).toSec() - time_init);
}

void update_map(const nav_msgs::OccupancyGrid::ConstPtr &map)
{
    mapdata = *map;
}

bool plan_path(nav_msgs::GetPlan::Request &plan_request, nav_msgs::GetPlan::Response &resp)
{
    int width = mapdata.info.width;
    int height = mapdata.info.height;
    float res = mapdata.info.resolution;
    float gox = mapdata.info.origin.position.x;
    float goy = mapdata.info.origin.position.y;
    Node start((int)floor((plan_request.start.pose.position.x - gox) / res),
               (int)floor((plan_request.start.pose.position.y - goy) / res), 0);
    Node goal((int)floor((plan_request.goal.pose.position.x - gox) / res),
              (int)floor((plan_request.goal.pose.position.y - goy) / res));
    std::vector<geometry_msgs::Point> path_of_points;
    frontier_vis.cells.clear();
    a_star(start, goal, path_of_points);
    nav_msgs::Path path;
    geometry_msgs::PoseStamped pose;
    path.header.frame_id = pose.header.frame_id = "/map";
    path.header.stamp = pose.header.stamp = ros::Time(0);
    pose.pose.orientation.w = 1.0;
    if (path_of_points.size() <= 3)
    {
        ROS_INFO("A* has not found a valid path within the alotted time. Send a new goal point.");
        path.poses.push_back(pose);
        pose.header.seq = 420;
        resp.plan = path;
        return true;
    }
    BOOST_FOREACH (geometry_msgs::Point &point, path_of_points)
    {
        pose.pose.position.x = point.x;
        pose.pose.position.y = point.y;
        path.poses.push_back(pose);
    }
    resp.plan = path;
    ROS_INFO("A* has computed the path");
    return true;
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "mnlc_a_star");
    ros::NodeHandle n;
    n.getParam("/a_star/start_time", start_time);
    n.getParam("/controller/timeout", timeout);
    n.getParam("/a_star/obstacle_cost", obstacle_cost);
    nav_msgs::OccupancyGrid global_costmap;
    tf::TransformListener listener;
    ros::Subscriber cspace_sub = n.subscribe("/mnlc_global_costmap_opencv/cspace", 1, update_map);
    ros::Publisher frontier_pub = n.advertise<nav_msgs::GridCells>("/a_star/frontier", 1);
    ros::Time st;
    st.fromSec(start_time);
    ros::Time::sleepUntil(st);
    ros::service::waitForService("/rtabmap/get_map", ros::Duration(timeout));
    ros::ServiceClient client1 = n.serviceClient<const nav_msgs::GetMap>("/rtabmap/get_map");
    nav_msgs::GetMap srv1;
    if (client1.call(srv1))
    {
        ROS_INFO("RTabMap Mapping service call successful.");
    }
    else
    {
        ROS_ERROR("RTabMap Mapping service call failed.");
        exit(0);
    }
    if (srv1.response.map.info.resolution == 0)
    {
        ROS_ERROR("RTabMap has zero resolution. Exiting...");
        exit(0);
    }
    nav_msgs::OccupancyGrid initial_map_metadata = srv1.response.map;
    ros::Time sr;
    sr.fromSec(timeout);
    ros::Time::sleepUntil(sr);
    ros::spinOnce();
    ros::service::waitForService("/begin_phase1", ros::Duration(timeout));
    ros::ServiceClient client2 = n.serviceClient<const std_srvs::Empty>("/begin_phase1");
    std_srvs::Empty srv2;
    if (client2.call(srv2))
    {
        ROS_INFO("Begin phase1 service call successful.");
    }
    else
    {
        ROS_ERROR("Begin phase1 service call failed:");
        exit(0);
    }
    listener.waitForTransform("/odom", "/base_footprint", ros::Time(0), ros::Duration(timeout));
    ros::ServiceServer planner = n.advertiseService("/plan_path", plan_path);
    frontier_vis.header.frame_id = "/map";
    frontier_vis.header.stamp = ros::Time(0);
    ros::Rate loop_rate(120);
    while (ros::ok())
    {
        ros::spinOnce();
        frontier_pub.publish(frontier_vis);
        loop_rate.sleep();
    }
    return 0;
}